containertest/docker.go

```diff
@@ -0,0 +1,175 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package containertest
+
+// This file implements docker integration.
+//
+// This file is only intended to be used outside of Google. Inside of Google, this file should be
+// replaced with the Google-internal version.
+
+import (
+	"fmt"
+	"io"
+	"strings"
+	"sync"
+	"time"
+
+	dockertest "github.com/ory/dockertest/v3"
+	"github.com/ory/dockertest/v3/docker"
+)
+
+var nopCloser = io.NopCloser(nil)
+
+// start starts a docker container running a service. A struct is returned describing how to
+// connect to the container, along with a cleanup function that should be called once all tests have
+// finished.
+//
+// The returned ConnInfo.Close() should be called in every case, even if this function returns an error. This
+// ensures that the Docker container will be cleaned up if the error occurred after the container
+// was created. The ConnInfo.Close() will never be nil.
+//
+// (For databases): Since the startup time for database containers can be as long as 20 seconds,
+// we share the container among every test. Each test should use a randomly-created
+// database/schema name to avoid collisions between tests.
+//
+// This function installs a signal handler for SIGTERM, SIGKILL, and SIGINT that attempts to clean
+// up the Docker container, then runs os.Exit(1). Since the signal handler kills the process, any
+// other custom signal handlers that are installed may not get a chance to run.
+//
+// Docker must be installed on localhost for this to work. No environment vars are needed.
+func start(conf *config) (ConnInfo, error) {
```

Comment:
This should return a pointer to a ConnInfo (same with startContainer)

Reply:
OK, I had made it the concrete type as a hint that it would never be `nil`, is the main consideration just that passing structs by value is too expensive if they grow to be large?

Reply:
Should MustStart return a pointer too? Or not because its an interface boundary?

Reply:
Changed to pointer